File: brokenCalc.java
// **Optimized → Go from target to → startvalue (Ask interview if you are allowed to use /2 and +1 i.e division and addition operation) take greedy path i.e best path**
//     1. If number is divisible by 2 → then only divide it or add 1 to it
//     2. Once your number reaches less than target → target - number = number of steps (i.e you need to add 1 these many times, so add this number to your step count)
//     3. TC: log(target value) SC: O(1)
class Solution {
    public int brokenCalc(int startValue, int target) {
        int steps = 0;
        
        while(target > startValue){
            if(target % 2 == 0){
                target = target/2;
            }else{
                target = target + 1;
            }
            steps++;
        }
        return steps + (startValue - target);
    }
}

File: broken_calc.py
# Time Complexity: O(log(target))
# Space Complxty: O(target/2)

class Solution:
    def brokenCalc(self, startValue: int, target: int) -> int:
        count = 0
        while target > startValue:
            if target % 2 == 0:
                target = target//2
            else:
                target += 1
            count += 1
        return count + (startValue - target)

        # q = deque()
        # if startValue == target: return 0
        # q.append(startValue - 1)
        # q.append(startValue*2)
        # count = 1
        # while q:
        #     size = len(q)
        #     for i in range(size):
        #         x = q.popleft()
        #         if x == target: return count
        #         q.append(x-1)
        #         if x < target:q.append(x*2)
        #     count += 1
        # return count
